nixos/parsedmarc: Fix build and runtime

e1mo opened the pull request:
## Description of changes

Currently the `parsedmarc` Module is not working. This is due to:
1. `smtp.to` having defaults causing evaluation errors.
2. The default `settings.elasticsearch.hosts` (if elasticsearch is to be provisioned) not being a valid URL, causing errors.
3. `python311Packages.parsedmarc` not building successfully due to missing dependencies / incompatible versions of some dependencies.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

mweinelt commented on pkgs/development/python-modules/parsedmarc/default.nix:
```diff
@@ -51,7 +51,8 @@ buildPythonPackage rec {
 
   postPatch = ''
     substituteInPlace pyproject.toml \
-      --replace "elasticsearch<7.14.0" "elasticsearch"
+      --replace "elasticsearch<7.14.0" "elasticsearch" \
+      --replace "elasticsearch-dsl==7.4.0" "elasticsearch-dsl"
   '';
```
Use `pythonRelaxDepsHook` instead.

```nix
  nativeBuildInputs = [
    pythonRelaxDepsHook
  ];

  pythonRelaxDeps = [
    "elasticseearch"
    "elasticsearch-dsl"
  ];

mweinelt reviewed (dismissed):
Please update the attribute for the python-module changes to be prefixed with `python311Packages`, since `pythonPackages` is a reference for `python2.pkgs`.

mweinelt commented:
~1400 rebuilds on each -linux, ~600 on each -darwin. Likely a result from touching the elasticsearch module.

But going through a staging cycle to get the propagation fixed seems unreasonable.

mweinelt commented on pkgs/development/python-modules/elasticsearch/default.nix:
```diff
@@ -21,13 +21,10 @@ buildPythonPackage rec {
-  ];
-
   propagatedBuildInputs = [
     urllib3
     certifi
```
```suggestion
```

https://github.com/elastic/elasticsearch-py/blob/v8.9.0/setup.py#L54

mweinelt commented:
Are the package changes also required for release-23.11? If not, could we only backport the module fixes, since they cause no rebuild?

e1mo commented:
The backporting of the package fix is needed indeed :/

Parsedmarc tests on `release-23.11` do also fail with `ModuleNotFoundError: No module named 'elastic_transport'`.

Janik-Haag commented:
more or less a duplicate of #279068

dotlambda committed:
python311Packages.parsedmarc: mark broken

dotlambda committed:
parsedmarc: override msgraph-core version

dotlambda committed:
nixos/parsedmarc: use top-level parsedmarc package

e1mo committed:
python311Packages.parsedmarc: Add missing dependency

e1mo committed:
python311Packages.parsedmarc: switch to pythonRelaxDepsHook and remove unneeded dependency

Co-Authored-By: Martin Weinelt <hexa@darmstadt.ccc.de>

e1mo committed:
nixos/parsedmarc: Fix smtp.to and default elasticsearch url

Previously evaluaton would fail if `smtp.to` was not set, since the
default case was not handled.

With the current versions of the python elasticsearch libraries, any
configuration containing the modules default would fail since it's not a
valid URL.

The issue adressed in #279068 is also adressed, thuse closes #279068.

e1mo commented:
I did bring this PR up to the current state of the module & package.
The latest update introduced `opensearch-py` new dependency which was not accounted for (but now is). The `msgraph-core` version is now pinned to `0.2.2`, thus parsedmarc is building again.

Since I would kinda like to have a working parsedmarc package and module again, I've kept the `smtp.to` patches that are a duplicate of #279068, as mentioned by @Janik-Haag.
Because those fixes on their own would not fix the broken build (and thus not fix broken deployments), I would prefer to go with this PR fixing all of breaking issues at once.

And maybe #285840 could get some fresh momentum to help get parsedmarc out of the persistent state of breakage?

mweinelt commented:
Unfortunately overrides of libraries aren't possible within the python package set. That would only be possible if `parsedmarc` was built with `buildPythonApplication` outside `python3Packages`.

The reason is that libraries propagate their dependencies, and they ultimately all end up in a single PYTHONPATH that only supports exactly one version of a python module. With overrides in libraries it would be too easy to end up with multiple different propagated module versions, which would make the package set unmaintainable.

e1mo commented:
Thanks for the explanation! I got kinda fooled by all tests passing (since we're not testing anything msgraph this makes sense tho) :/

Do you have any concerns regarding the changes introduced in #283337 (especially edf24111e45a1cc110fae47b9a8905b67e0983d7)? If not, I'd use them here too.

mweinelt commented:
Yeah, that looks plausible.

SuperSandro2000 reviewed (approved):
LGTM

e1mo commented:
I can't come up with anything that is still left open.
I also re-ran both parsedmarc test locally with the recent rebase and am happy to report that they're still passing :D
